JAX-RS build fails with UnsupportedOperationException

### Describe the bug

We are upgrading a Quarkus 2.16.12 app to 3.15.1.  We are also replacing our legacy swagger jars to generate the OpenApi files using the OpenAPI microprofile approach.

One of the differences we found is that if we put @JsonView(Views.Ingest.class) at the method level it only was applied to the GET operations.  For POST & PUT operations we had to move the @JsonView(Views.Ingest.class) to the RequestBody entity parameter.  Without placing the annotation at the request body parameter it would always use the full entity spec instead of the one defined by @JsonView(Views.Ingest.class).

However now that I have made this change for all of our REST services I get the following error.  What is causing this?  How can I fix this?

I am not using @CustomSerialization or @CustomDeserialization` in my code but Quarkus might be using these at build time.


[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:3.15.1:build (quarkus-maven-plugin-build) on project sdl: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleJsonAnnotations threw an exception: java.lang.UnsupportedOperationException: The `@CustomSerialization` and `@CustomDeserialization` annotations can only be used in methods or classes.
[ERROR]         at io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor.getTargetId(ResteasyReactiveJacksonProcessor.java:631)
[ERROR]         at io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor.handleJsonAnnotations(ResteasyReactiveJacksonProcessor.java:268)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:483)

Thanks,
David

### Expected behavior

Build would be successful.

### Actual behavior

Generates the build error above.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

java 21.0.4 2024-07-16 LTS

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java
@@ -75,7 +75,11 @@ import io.quarkus.resteasy.reactive.jackson.runtime.ResteasyReactiveServerJackso
 import io.quarkus.resteasy.reactive.jackson.runtime.mappers.NativeInvalidDefinitionExceptionMapper;
 import io.quarkus.resteasy.reactive.jackson.runtime.security.RolesAllowedConfigExpStorage;
 import io.quarkus.resteasy.reactive.jackson.runtime.security.SecurityCustomSerialization;
-import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.*;
+import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.BasicServerJacksonMessageBodyWriter;
+import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyReader;
+import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter;
+import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.GeneratedSerializersRegister;
+import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.ServerJacksonMessageBodyReader;
 import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.vertx.VertxJsonArrayMessageBodyReader;
 import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.vertx.VertxJsonArrayMessageBodyWriter;
 import io.quarkus.resteasy.reactive.jackson.runtime.serialisers.vertx.VertxJsonObjectMessageBodyReader;
@@ -263,7 +267,7 @@ public class ResteasyReactiveJacksonProcessor {
                     if ((jsonViews == null) || (jsonViews.length == 0)) {
                         continue;
                     }
-                    recorder.recordJsonView(getTargetId(instance.target()), jsonViews[0].name().toString());
+                    recorder.recordJsonView(getTargetId(instance), jsonViews[0].name().toString());
                 }
             }
             if (resourceClass.annotationsMap().containsKey(CUSTOM_SERIALIZATION)) {
@@ -290,7 +294,7 @@ public class ResteasyReactiveJacksonProcessor {
                             ReflectiveClassBuildItem.builder(biFunctionType.name().toString())
                                     .reason(getClass().getName())
                                     .build());
-                    recorder.recordCustomSerialization(getTargetId(instance.target()), biFunctionType.name().toString());
+                    recorder.recordCustomSerialization(getTargetId(instance), biFunctionType.name().toString());
                 }
             }
             if (resourceClass.annotationsMap().containsKey(CUSTOM_DESERIALIZATION)) {
@@ -317,7 +321,7 @@ public class ResteasyReactiveJacksonProcessor {
                             ReflectiveClassBuildItem.builder(biFunctionType.name().toString())
                                     .reason(getClass().getName())
                                     .build());
-                    recorder.recordCustomDeserialization(getTargetId(instance.target()), biFunctionType.name().toString());
+                    recorder.recordCustomDeserialization(getTargetId(instance), biFunctionType.name().toString());
                 }
             }
         }
@@ -641,15 +645,18 @@ public class ResteasyReactiveJacksonProcessor {
         return false;
     }
 
-    private String getTargetId(AnnotationTarget target) {
+    private String getTargetId(AnnotationInstance instance) {
+        AnnotationTarget target = instance.target();
         if (target.kind() == AnnotationTarget.Kind.CLASS) {
             return getClassId(target.asClass());
         } else if (target.kind() == AnnotationTarget.Kind.METHOD) {
             return getMethodId(target.asMethod());
+        } else if (target.kind() == AnnotationTarget.Kind.METHOD_PARAMETER) {
+            return getMethodId(target.asMethodParameter().method());
         }
 
-        throw new UnsupportedOperationException("The `@CustomSerialization` and `@CustomDeserialization` annotations can only "
-                + "be used in methods or classes.");
+        throw new UnsupportedOperationException(String.format("The `%s` annotation can only "
+                + "be used in methods or classes.", instance.name()));
     }
 
     private String getClassId(ClassInfo classInfo) {

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/JsonViewOnMethodParameterTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/JsonViewOnMethodParameterTest.java
@@ -1,0 +1,64 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test;
+
+import static io.restassured.RestAssured.given;
+import static jakarta.ws.rs.core.Response.Status.CREATED;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+
+import java.util.function.Supplier;
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.RestResponse;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.fasterxml.jackson.annotation.JsonView;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class JsonViewOnMethodParameterTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(User.class, Views.class, Resource.class);
+                }
+            });
+
+    @Test
+    public void test() {
+        given().accept("application/json")
+                .contentType("application/json")
+                .body("""
+                        {
+                         "id": 1,
+                         "name": "Foo"
+                        }
+                        """)
+                .post("test")
+                .then()
+                .statusCode(201)
+                .body(not(containsString("1")), containsString("Foo"));
+    }
+
+    @Path("test")
+    public static class Resource {
+
+        @POST
+        @Produces(MediaType.APPLICATION_JSON)
+        @Consumes(MediaType.APPLICATION_JSON)
+        public RestResponse<User> create(@JsonView(Views.Public.class) User user) {
+            return RestResponse.status(CREATED, user);
+        }
+    }
+}

